crates/driver/src/tests/cases/protocol_fees.rs

```diff
@@ -549,3 +729,291 @@ async fn price_improvement_fee_sell_out_of_market_order_capped() {
 }
+
+#[tokio::test]
+#[ignore]
+async fn price_improvement_fee_partial_buy_in_market_order_not_capped() {
+    let fee_policy = Policy::PriceImprovement {
+        factor: 0.5,
+        // high enough so we don't get capped by volume fee
+        max_volume_factor: 1.0,
+        quote: Quote {
+            sell: 39.ether().into_wei(),
+            buy: 40.ether().into_wei(),
+            network_fee: 1.ether().into_wei(),
+        },
+    };
+    let test_case = TestCase {
+        fee_policy,
+        order: Order {
+            // Demanding to sell more than quoted (in-market)
+            sell_amount: 50.ether().into_wei(),
+            buy_amount: 40.ether().into_wei(),
+            side: order::Side::Buy,
+        },
+        execution: Execution {
+            // Sell 10 ETH less than quoted, half of which is kept by the protocol
+            solver: Amounts {
+                sell: 10.ether().into_wei(),
+                buy: 20.ether().into_wei(),
+            },
+            driver: Amounts {
+                sell: 15.ether().into_wei(),
+                buy: 20.ether().into_wei(),
+            },
+        },
+    };
+    protocol_fee_test_case(test_case).await;
+}
+
+#[tokio::test]
+#[ignore]
+async fn price_improvement_fee_partial_sell_in_market_order_not_capped() {
+    let fee_policy = Policy::PriceImprovement {
+        factor: 0.5,
+        // high enough so we don't get capped by volume fee
+        max_volume_factor: 1.0,
+        quote: Quote {
+            sell: 49.ether().into_wei(),
+            buy: 50.ether().into_wei(),
+            network_fee: 1.ether().into_wei(),
+        },
+    };
+    let test_case = TestCase {
+        fee_policy,
+        order: Order {
+            sell_amount: 50.ether().into_wei(),
+            // Demanding to receive less than quoted (in-market)
+            buy_amount: 40.ether().into_wei(),
+            side: order::Side::Sell,
+        },
+        execution: Execution {
+            // Receive 10 ETH more than quoted, half of which gets captured by the protocol
+            solver: Amounts {
+                sell: 20.ether().into_wei(),
+                buy: 30.ether().into_wei(),
+            },
+            driver: Amounts {
+                sell: 20.ether().into_wei(),
+                buy: 25.ether().into_wei(),
+            },
+        },
+    };
+    protocol_fee_test_case(test_case).await;
+}
+
+#[tokio::test]
+#[ignore]
+async fn price_improvement_fee_partial_buy_out_of_market_order_not_capped() {
+    let fee_policy = Policy::PriceImprovement {
+        factor: 0.5,
+        // high enough so we don't get capped by volume fee
+        max_volume_factor: 1.0,
+        quote: Quote {
+            sell: 59.ether().into_wei(),
+            buy: 50.ether().into_wei(),
+            network_fee: 1.ether().into_wei(),
+        },
+    };
+    let test_case = TestCase {
+        fee_policy,
+        order: Order {
+            // Demanding to sell less than quoted (out-market)
```

Comment:
Do we expect a price improvement fee to ever apply to an out-of-market order?
In this case we are falling back to taking a fee compared to the limit price, right?

Reply:
In this case price improvement behaves exactly like Surplus fees (assuming the same factor). Depending on the final fee policy that is decided we may be able to use a single fee policy for all order types.